api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto

```diff
@@ -92,6 +93,9 @@ message ThriftProxy {
 
   // Optional maximum requests for a single downstream connection. If not specified, there is no limit.
   google.protobuf.UInt32Value max_requests_per_connection = 7;
+
+  // Use xDS to fetch the route configuration. If both route_config and rds defined, rds has priority.
+  Rds rds = 8;
```

Comment:
Probably want to use a oneof migration annotation here, so that we know to put this field and the `route_config` field in a oneof in the next major version (if that ever happens).

Reply:
Sorry, it is not clear. What do you mean 'it's not obvious to me that we want to use exactly the same resource type for HTTP RouteConfiguration and Thrift RouteConfiguration'? And 'two resource types happen to be the same right now, who's to say that they won't diverge in the future'?

As XDS resource the HTTP RouteConfiguration and Thrift RouteConfiguration are different for the two different protocols and cannot be exchanged.
But technically the GRPC endpoint RouteDiscoveryService can be used to transfer both. And they are both routing config so we don't try to force something unrelated into it.
I just meant technically it is not needed to introduce new endpoint like let's say RouteDiscoveryServiceThrift for Thrift RouteConfiguration. In type_to_endpoint.cc the endpoint is selected based on the resource type_url with that assumption there is 1:1 mapping between endpoints and resource urls. With this 'alias' we can say: ok use the same endpoint to this resource type like that resource type.  Maybe not the most beautiful but simple and no need to touch this part again if we introduce support for RDS for other protocols.

Yes, in the first version I used 'oneof'. But then that extreme but previously valid config where neither 'route_config' nor 'rds' specified becomes invalid.

Reply:
> What do you mean 'it's not obvious to me that we want to use exactly the same resource type for HTTP RouteConfiguration and Thrift RouteConfiguration'? And 'two resource types happen to be the same right now, who's to say that they won't diverge in the future'?

What I'm saying is that the resource type `envoy.config.route.v3.RouteConfiguration` was designed to contain routing information for HTTP specifically.  Even if thrift happens to need exactly the same format today, it's not clear to me that it necessarily makes sense to use the same resource type directly here, because we don't want to be in a position where in the future we need to either add a bunch of thrift-specific fields to the existing `envoy.config.route.v3.RouteConfiguration` proto or we need to break thrift routing by changing what resource type it uses after everyone is already expecting to use the current resource type.

Alternatively, though, if we can migrate the existing `envoy.config.route.v3.RouteConfiguration` proto to be a more generic form (possibly via the work that @snowp is doing in #17633), then our story can be that that resource type is not actually HTTP-specific, and we have enough extensibility to make it work for any resource type.  I'd like @mattklein123 and @snowp to weigh in on that option.

I think the discussion about the RPC service (use the existing `RouteDiscoveryService` vs. a new one) is actually just a side question; the main issue is what resource type we use here.  The resource type generally maps 1:1 to the RPC service, so if we go with the same resource type, then it makes sense to use the same RPC service, and if we go with a different resource type, then it makes sense to add a new RPC service.

> Yes, in the first version I used 'oneof'. But then that extreme but previously valid config where no 'route_config' specified becomes invalid.

Using a `oneof` does not make the fields in the `oneof` required.  If both `route_config` and `rds` were in the same `oneof`, it would still be legal to specify neither of them.

Reply:
@tkovacs-2 the Envoy-side code changes are great, thanks for the refactoring, it's much cleaner.

From what I can see above, we actually have two distinct resources for Thrift and Envoy RDS, we're not reusing `RouteConfiguration` here. I am a bit confused about the "resource alias" annotation though, that seems to imply some connection which Mark points out doesn't have any reasonable association in the API.